File: frontend/src/types/report.ts
import type { UserRole } from './auth';
import type { ApprovalDecision, TaskPriority, TaskStatus } from './task';

export interface StatusCount {
  status: TaskStatus;
  count: number;
}

export interface PriorityCount {
  priority: TaskPriority;
  count: number;
}

export interface UserWorkloadEntry {
  user_id: number;
  name: string;
  email: string;
  role: UserRole;
  total: number;
  pending: number;
  in_progress: number;
  submitted: number;
  approved: number;
  rejected: number;
  completed: number;
}

export interface ApprovalSummaryEntry {
  decision: ApprovalDecision;
  count: number;
}

export interface TasksByStatusReport {
  items: StatusCount[];
  total: number;
}

export interface TasksByPriorityReport {
  items: PriorityCount[];
  total: number;
}

export interface UserWorkloadReport {
  items: UserWorkloadEntry[];
}

export interface ApprovalSummaryReport {
  items: ApprovalSummaryEntry[];
  total: number;
}

File: backend/app/schemas/report.py
"""Reporting schemas."""
from __future__ import annotations

from typing import List

from pydantic import BaseModel

from app.models.approval import ApprovalDecision
from app.models.task import TaskPriority, TaskStatus
from app.models.user import UserRole


class StatusCount(BaseModel):
    status: TaskStatus
    count: int


class PriorityCount(BaseModel):
    priority: TaskPriority
    count: int


class UserWorkloadEntry(BaseModel):
    user_id: int
    name: str
    email: str
    role: UserRole
    total: int
    pending: int
    in_progress: int
    submitted: int
    approved: int
    rejected: int
    completed: int


class ApprovalSummaryEntry(BaseModel):
    decision: ApprovalDecision
    count: int


class TasksByStatusReport(BaseModel):
    items: List[StatusCount]
    total: int


class TasksByPriorityReport(BaseModel):
    items: List[PriorityCount]
    total: int


class UserWorkloadReport(BaseModel):
    items: List[UserWorkloadEntry]


class ApprovalSummaryReport(BaseModel):
    items: List[ApprovalSummaryEntry]
    total: int
